modules/gapi/test/gapi_transform_tests.cpp

```diff
@@ -146,44 +164,102 @@ TEST(KernelPackageTransform, gmat_gsc_in_gmat_out)
     check(tr.substitute());
 }
 
-TEST(KernelPackageTransform, gmat_in_garr_out)
+namespace
 {
-    auto tr = gmat_in_garr_out::transformation();
+    template<typename InType, typename OutType>
+    typename std::enable_if<(cv::detail::GTypeTraits<OutType>::kind == cv::detail::ArgKind::GARRAY), void>::type
+    array_check(const cv::GComputation::Priv &p)
+    {
+        EXPECT_TRUE(ProtoContainsT<InType>(p.m_ins[0]));
 
-    auto check = [](const cv::GComputation &comp){
+        EXPECT_TRUE(ProtoContainsT<cv::detail::GArrayU>(p.m_outs[0]));
+        EXPECT_TRUE(cv::util::get<cv::detail::GArrayU>(p.m_outs[0]).holds<ArrayT>());
+        EXPECT_FALSE(cv::util::get<cv::detail::GArrayU>(p.m_outs[0]).holds<WrongArrayT>());
+    }
+
+    template<typename InType, typename OutType>
+    typename std::enable_if<(cv::detail::GTypeTraits<InType>::kind == cv::detail::ArgKind::GARRAY), void>::type
+    array_check(const cv::GComputation::Priv &p)
+    {
+        EXPECT_TRUE(ProtoContainsT<OutType>(p.m_outs[0]));
+
+        EXPECT_TRUE(ProtoContainsT<cv::detail::GArrayU>(p.m_ins[0]));
+        EXPECT_TRUE(cv::util::get<cv::detail::GArrayU>(p.m_ins[0]).holds<ArrayT>());
+        EXPECT_FALSE(cv::util::get<cv::detail::GArrayU>(p.m_ins[0]).holds<WrongArrayT>());
+    }
+
+    template<typename InType, typename OutType>
+    typename std::enable_if<(cv::detail::GTypeTraits<InType>::kind == cv::detail::ArgKind::GARRAY ||
+                             cv::detail::GTypeTraits<OutType>::kind == cv::detail::ArgKind::GARRAY), void>::type
+    args_check(const cv::GComputation &comp)
+    {
         const auto &p = comp.priv();
         EXPECT_EQ(1u, p.m_ins.size());
         EXPECT_EQ(1u, p.m_outs.size());
 
-        EXPECT_TRUE(ProtoContainsT<GMat>(p.m_ins[0]));
-
-        EXPECT_TRUE(ProtoContainsT<cv::detail::GArrayU>(p.m_outs[0]));
-        EXPECT_TRUE(cv::util::get<cv::detail::GArrayU>(p.m_outs[0]).holds<int>());
-        EXPECT_FALSE(cv::util::get<cv::detail::GArrayU>(p.m_outs[0]).holds<float>());
+        array_check<InType, OutType>(p);
     };
 
-    check(tr.pattern());
-    check(tr.substitute());
+    template<typename InType, typename OutType>
+    typename std::enable_if<(cv::detail::GTypeTraits<InType>::kind != cv::detail::ArgKind::GARRAY &&
+                             cv::detail::GTypeTraits<OutType>::kind != cv::detail::ArgKind::GARRAY), void>::type
+    args_check(const cv::GComputation &comp)
+    {
+        const auto &p = comp.priv();
+        EXPECT_EQ(1u, p.m_ins.size());
+        EXPECT_EQ(1u, p.m_outs.size());
+        EXPECT_TRUE(ProtoContainsT<InType>(p.m_ins[0]));
+        EXPECT_TRUE(ProtoContainsT<OutType>(p.m_outs[0]));
+    }
```

Comment:
Can all this checks be splitt into GArray / non-GArray check independent on input/output-ness? (So you can utilize such check for both input and output:)
```
template<typename Type>
typename std::enable_if<cv::detail::GTypeTraits<Type>::kind == cv::detail::ArgKind::GARRAY, void>::type
arg_check(const cv::GProtoArg &arg)
{
	// check array
}

// here should be a similar arg_check function for non-garray

// and args_check becomes something like
template<typename InType, typename OutType>   
args_check(const cv::GComputation &comp)
{
    const auto &p = comp.priv();
    EXPECT_EQ(1u, p.m_ins.size());
    EXPECT_EQ(1u, p.m_outs.size());

    arg_check<InType>(p.m_ins[0]);
    arg_check<OutType>(p.m_outs[0]);
}
```